pkg/cloud-controller-manager/loadBalancer.go

```diff
@@ -80,67 +93,24 @@ func (l *LoadBalancerManager) GetLoadBalancerName(ctx context.Context, clusterNa
 	return name
 }
 
-// EnsureLoadBalancer is to create/update a Harvester load balancer for the service and return the loadBalancerStatus with an IP
-// 1. watch the LB to get an IP asynchronously
-// 2. create or update lb
-// 3. wait for an ip and return the LoadBalancerStatus
+// EnsureLoadBalancer is to create/update a Harvester load balancer for the service
+// 1. Create/update harvester load balancer.
+//    If the service has an external IP set by kube-vip, update it into the load balancer.
+// 2. Wait for the harvester load balancer to get the allocated IP address
+// 3. Set the allocated IP address into the field spec.loadBalancerIP of the service.
+//    The kube-vip will set the external IP according to the spec.loadBalancerIP.
 func (l *LoadBalancerManager) EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error) {
 	name := l.GetLoadBalancerName(ctx, clusterName, service)
-	ipChan := make(chan string)
 
-	lb, getErr := l.lbClient.Get(l.namespace, name, metav1.GetOptions{})
-	if getErr != nil && !errors.IsNotFound(getErr) {
-		return nil, getErr
+	if err := l.createOrUpdateLoadBalancer(name, clusterName, service); err != nil {
+		return nil, fmt.Errorf("create or update lb %s/%s failed, error: %w", l.namespace, name, err)
 	}
 
-	// watch the lb to get an ip
-	w, err := l.lbClient.Watch(l.namespace, metav1.ListOptions{FieldSelector: fmt.Sprintf("metadata.name=%s", name)})
-	if err != nil {
-		return nil, fmt.Errorf("watch loadbalancer in namespace %s error, %w", l.namespace, err)
-	}
-	defer w.Stop()
-	go getIP(w, ipChan, lb)
-
-	// create or update lb
-	if getErr == nil {
-		if err := l.updateLoadBalancer(lb, service, nodes); err != nil {
-			return nil, err
-		}
-	} else {
-		spec, err := getLBSpec(service, nodes)
-		if err != nil {
-			return nil, err
-		}
-		if _, err := l.lbClient.Create(&lbv1.LoadBalancer{
-			ObjectMeta: metav1.ObjectMeta{
-				Namespace: l.namespace,
-				Name:      name,
-				Annotations: map[string]string{
-					uuidKey: string(service.UID),
-				},
-				Labels: map[string]string{
-					clusterNameKey: clusterName,
-				},
-			},
-			Spec: *spec,
-		}); err != nil {
-			return nil, err
-		}
+	if err := l.updateServiceLoadBalancerIP(name, service); err != nil {
+		return nil, fmt.Errorf("update load balancer IP of service %s/%s failed, error: %w", service.Namespace, service.Name, err)
```

Comment:
load balancer -->  loadbalancer
better also print l.Name,  l.IP  in error message